### DISCUSSION TITLE
Does rapidfuzz work with other languages than English?

### DISCUSSION THREAD

**Discussion by lingvisa** [rapidfuzz/RapidFuzz#203] [Q&A] [answered]
I am working with Asian languages like Japanese, Chinese. Does rapidfuzz work with non-ascill characters? 

Thanks for clarification.

**Comment by maxbachmann** [comment, marked as answer]
Yes RapidFuzz works with any unicode characters.

Note however that many of the implementations have an optimized implementation for characters below 256 (extended ascii), since it can directly look those up in an array[256] which is not possible for the full unicode range. So when you run into performance problems it might help to transform the most common characters of the language to ascii characters.

**Reply by lingvisa** [reply]
@maxbachmann Would those chars below 256 have already covered most typical Japanese and Chinese characters? If not, how should I do such a transformation?

**Reply by maxbachmann** [reply]
> Does rapidfuzz work with non-ascill characters?

Thinking about this again: All token based scorers probably do not work well with chinese (token_sort_ratio/token_set_ratio/token_ratio/partial_token_sort_ratio/partial_token_set_ratio/partial_token_ratio/WRatio), since they split the text into words at spaces. Since chinese does not use spaces this will likely not work. For others like the Levenshtein distance the results might not be great, since words consist of a lot less characters in chinese. So it might make sense to work with radicals/strokes instead: https://stackoverflow.com/a/12380762/11335032


> Would those chars below 256 have already covered most typical Japanese and Chinese characters

No only extended ascii is below 256

> If not, how should I do such a transformation?

Thinking about this again, I am not sure how well this works for chinese given the very large amount of characters. However the basic idea would be that you take the 256 most used characters of the language and replace those characters in your dataset with the charactes 0-255 (as long as those can not appear in your data).

Note that you should probably benchmark your dataset vs a pure ascii dataset of similar size to see whether the performance difference even matters in your use case. Especially for short strings (<64 characters) it likely will not matter very much, while for very long strings (e.g 100k characters) it will matter.

**Reply by lingvisa** [reply]
If I understand correctly, 256 characters are too few for Chinese words, while english has only a-z, 0-9 chars.  I thought no matter what language it is, those fuzzy or trie based matching use characters, not tokens. So why does it need tokenization?

**Reply by lingvisa** [reply]
Add a little more context: My use case is to given a piece of text, usually a post from social media, and  a large dictionary with terms to be matched against with the text, I want to efficiently identify exact matches and fuzzy matches with a high confidence score, i.e. 95%. So it's not one short phrase to compare with another short phrase.

**Reply by lingvisa** [reply]
I am looking at this link: https://github.com/jackmen/PhuzzyMatcher/blob/master/scripts/fuzzy_phrase_matcher.py, which is based on your rapidfuzz implementation. Do you think 'results = self.fuzzy_matcher(self.set, doc.text.lower(), self.match)' is efficient? Spacy's PhraseMatcher itself is efficient, but only supports exact matches.

**Reply by maxbachmann** [reply]
>  I thought no matter what language it is, those fuzzy or trie based matching use characters, not tokens. So why does it need tokenization?

The underlying algorithms: `fuzz.ratio`/`fuzz.partial_ratio`/`distance.Levenshtein.*`/`distance.Indel.*`/`distance.Hamming.*` are all using characters. Some of the scorers like `fuzz.token_sort_ratio` preprocess the strings e.g. by sorting the words in the two texts before matching them, which requires tokenization.

> : My use case is to given a piece of text, usually a post from social media, and a large dictionary with terms to be matched against with the text

what does large mean? 10k/100k/1M/... terms

> So it's not one short phrase to compare with another short phrase.

I guess the length largely depends on the platform. E.g. on twitter the average post length is somewhere between 30 and 40 (probably shorter for languages like chinese) which I would consider as short

> I am looking at this link: https://github.com/jackmen/PhuzzyMatcher/blob/master/scripts/fuzzy_phrase_matcher.py, which is based on your rapidfuzz implementation. Do you think 'results = self.fuzzy_matcher(self.set, doc.text.lower(), self.match)' is efficient?

the fuzzy_matcher performs the following operation `if fuzz.ratio(matched_phrase, feature.lower()) > match:`
1) it would be more efficient to provide rapidfuzz with the score_cutoff
    `if fuzz.ratio(matched_phrase, feature.lower(), score_cutoff=match):`
2) it calls fuzz.ratio for each of the elements. This is inefficient, since each of the function calls adds over 100ns of overhead.

You can either use:
```python
process.extract(phrase, terms, scorer=fuzz.ratio, processor=None, score_cutoff=95, limit=None)
```
or
```python
process.cdist([phrase], terms, scorer=fuzz.ratio, score_cutoff=95)
```
instead which can avoid this overhead

**Reply by lingvisa** [reply]
By 'large', I mean around 1M and currently my terms are 800,000. My text length cutoff is 400, and cut it up to 400 chars if it more than that. For scoring it is using token, and if I don't want to use tokenization, I can set "processor=None", as you show, right?

**Reply by maxbachmann** [reply]
> For scoring it is using token, and if I don't want to use tokenization, I can set "processor=None", as you show, right?

the processor function allows additional preprocessing like lowercasing strings before matching them. If you use a scorer like `fuzz.ratio` no tokenization takes place. This only occurs for scorers like `fuzz.token_sort_ratio`. E.g. `fuzz.token_sort_ratio` tokenizes the texts, sorts it and then runs `fuzz.ratio` on the sorted text.

> By 'large', I mean around 1M and currently my terms are 800,000. My text length cutoff is 400

Do I understand it correctly that you would like to search for 1M texts in the 800k term database and each of the texts/terms can be up to 400 characters long?
I performed a quick benchmark on my laptop to search for a single random text inside an 800k database without multithreading:
ascii + all lengths 64 = 0.1s
non-ascii + all lengths 64 = 0.36s
ascii + all lengths 400 = 2.2s
non-ascii + all lengths 400 = 14.5s

So non-ascii appears to cause a 4-8x performance hit, which is pretty painful.
When all texts have a length of 400 this is pretty unpractical at least on my machine, since it would take around 168 days to search for the 1M texts. Even with  a faster CPU + multithreading this would still take quite a while. When a large amount of the texts is shorter or the are large length differences between the strings this will be faster, since it can often skip the similarity calculation given your high score_cutoff of 95. However performing an O(N/64 *M) algorithm 8e+11 times will always take a long time.

**Reply by lingvisa** [reply]
Sorry, actually, I meant that I currently have 800K terms, and it could be increased to 1m over time. So let's say the max count of dictionary term is 1m. Then, I process each piece of text, whose length could be <=400 chars (mainly Chinese characters). Any if there is any phrase that matches terms in the 1M terms, then return the phrases and their indexes. So, I think we just need to benchmark given each post, how much time it would spend to fuzzy match the dictionary terms, with a confidence of 95%.

**Reply by lingvisa** [reply]
If you mean it takes 14.5 seconds to process (fuzzy match) 1 million texts of 400 characters long, that would be very efficient, because on average it takes (14.5 * 1000)/1,000,000 ms to process one post. That seems to be too fast. Is that true?

**Reply by maxbachmann** [reply]
> Sorry, actually, I meant that I currently have 800K terms, and it could be increased to 1m over time.

That makes this a lot less problematic :)

> If you mean it takes 14.5 seconds to process (fuzzy match) 1 million texts of 400 characters long

It took 14.5 seconds to compare a single text to 800k texts. So around 18us per text comparison.
```python
def get_random_char():
    # random char outside of ascii range
    return chr(ord(random.choice(string.ascii_letters + string.digits)) + 1000)

def create_random_string(length):
    return ''.join(get_random_char() for _ in range(length))

terms = [create_random_string(400) for _ in range(800 * 1000)]
text = create_random_string(400)
process.extract(text, terms, scorer=fuzz.ratio, processor=None, score_cutoff=95, limit=None)
```

**Reply by lingvisa** [reply]
If that's case, it would be too slow in practice. For each processing of one post, it has to be in a few mill-seconds, to make it practical. Probably, I have to look for something else. Why is the exact match in Trie so efficient? I am currently using PhraseMatcher from spacy which does exact match with my big dictionary of 800K terms, and it's really too fast. Any other possibilities to make this fuzzy matching closer to PhraseMatcher?

**Reply by maxbachmann** [reply]
> Why is the exact match in Trie so efficient?

In a trie it only requires a single pass, which makes it O(N) to check whether there is a similar string stored in the trie. While for the fuzzy match it needs to run an O(N/64 *M) algorithm N times which results in cubic runtime. There are some other algorithms which can make use e.g. of the triangle inequality to improve the performance, but those are not part of rapidfuzz at this point.

**Reply by lingvisa** [reply]
In PhraseMatcher, usually one post takes only 1 ms to complete. Can you give me a rough idea of the complexity of 'triangle inequality' and how far away it is from trie match? In my application, I can not afford to spend  much time on dictionary match.

**Reply by maxbachmann** [reply]
One example of an alternative are levenshtein automatons as described e.g. here: https://blog.mikemccandless.com/2011/03/lucenes-fuzzyquery-is-100-times-faster.html

However I have absolutely no clue how fast those would be in your use case.

**Reply by lingvisa** [reply]
A second thought for optimization is that, my dictionary size can be cut into half by limiting the term length >=6 chars, because short sequence is much more ambiguous so the precision would be very low. By this way, the 800K reduces to 400K. Could you benchmark again how long does it take to process one post using the code above? The reduced time should be more than half. Right?

**Reply by maxbachmann** [reply]
> Could you benchmark again how long does it take to process one post using the code above?

In practice it will be a lot faster:
1) you have very different string lengths. RapidFuzz will skip many of those, since given your score_cutoff they can not be similar enough
2) shorter strings are faster to calculate

E.g. by selecting a random string length between 0 and 400 instead of making all strings length 400 I reach a lot faster run time. Here are the times for 1m elements with len 0-400
ascii        = 233us -> 2.3ns per element
non-ascii = 277us -> 2.8ns per element

This should be really benchmarked with your actual data, since this largely depends on the string lengths your dealing with. You should be able to easily benchmark this yourself with `process.extract`.

**Reply by lingvisa** [reply]
Yes, I will definitely test it with my data in integration with Spacy and update here once I do it.
